test/unit/puppetlabs/services/jruby/jruby_agents_test.clj

```diff
@@ -104,29 +105,28 @@
 
 (deftest next-instance-id-test
   (let [pool-context (jruby-core/create-pool-context
-                       (jruby-testutils/jruby-config {:max-active-instances 8})
-                       jruby-testutils/default-shutdown-fn)]
+                      (jruby-testutils/jruby-config {:max-active-instances 8}))]
     (testing "next instance id should be based on the pool size"
       (is (= 10 (jruby-agents/next-instance-id 2 pool-context)))
       (is (= 100 (jruby-agents/next-instance-id 92 pool-context))))
     (testing "next instance id should wrap after max int"
       (let [id (- Integer/MAX_VALUE 1)]
         (is (= (mod id 8) (jruby-agents/next-instance-id id pool-context)))))))
 
-;; TODO: this test is useless as-is; we need to add a cleanup callback and change
-;; this test to validate that it gets called.
-(deftest master-termination-test
-  (testing "Flushing the pool causes masters to be terminated"
+(deftest custom-termination-test
+  (testing "Flushing the pool causes shutdown hook to be called"
     (logutils/with-test-logging
-      (tk-testutils/with-app-with-config
-        app
-        [jruby/jruby-pooled-service]
-        (-> (jruby-testutils/jruby-tk-config
-             (jruby-testutils/jruby-config {:max-active-instances 1})))
-        (let [jruby-service (tk-app/get-service app :JRubyService)
-              context (tk-services/service-context jruby-service)]
-          (jruby-protocol/flush-jruby-pool! jruby-service)
-          ; wait until the flush is complete
-          (await (get-in context [:pool-context :pool-agent]))
-          ;; TODO: add termination callback, test that it is called?
-          #_(is (logged? #"Terminating Master")))))))
+     (let [config (assoc-in (jruby-testutils/jruby-tk-config
+                             (jruby-testutils/jruby-config {:max-active-instances 1}))
+                            [:jruby :lifecycle]
+                            {:shutdown (fn [x] (log/error "Hello from shutdown") x)})]
+        (tk-testutils/with-app-with-config
+         app
+         [jruby/jruby-pooled-service]
+         config
+         (let [jruby-service (tk-app/get-service app :JRubyService)
+               context (tk-services/service-context jruby-service)]
+           (jruby-protocol/flush-jruby-pool! jruby-service)
+           ; wait until the flush is complete
+           (await (get-in context [:pool-context :pool-agent]))
+           (is (logged? #"Hello from shutdown"))))))))
```

Comment:
Hmmm... I guess this is redundant? I feel like in the future, assuming we really do remove the TK service, it would be good to change this to not rely on a tk service at all, just call the underlying functions, and then maybe the one in the `int` namespace would define its own service and use that? i.e. I'm thinking that ultimately this test would be a bit more unit-y and the one in the `int` namespace would be more integration-y. Don't know if that's actually worthwhile though.